### ISSUE TITLE
Options missing for specifying secure private registry SSL certificate, etc. 

### ISSUE THREAD

**Issue by CalvinHartwell** [juju-solutions/layer-docker#113]
Hi all,

This docker layer doesn't seem to handle the options for secure private registries such as the SSL certificate, CA certificate, client certificate, etc. 

This means we can only handle insecure registries out of the box (using HTTP or HTTPS but ignoring the certificate). 

Cheers

**Comment by marcoceppi** [comment]
Do you mean secure registries where the CA for that secure registry is not on the machine? If not, and it's just a secure registry, there's no additional changes, just set the image you're using to `secure-registry.tld/user/image`

**Comment by CalvinHartwell** [comment]
> Do you mean secure registries where the CA for that secure registry is not on the machine? If not, and it's just a secure registry, there's no additional changes, just set the image you're using to secure-registry.tld/user/image

@marcoceppi sorry, yes for being vague, we were rushed out of the office as end of day - if we use a secured registry (I.E its serving over HTTPS) there seems to be no parameters to specify the public CA certificate (or chain of certificates) used to sign the public certificate of the secured registry. This is important as the client needs to be told what it can trust/not trust. 

Obviously this is not on the machine at deployment time, so surely the handshake will fail and Docker won't be able to pull the images from registry? Unless your saying docker just accepts ALL SSL certificates presented from a registry, meaning it doesn't really guarantee security at all.

**Comment by wwwtyro** [comment]
Just want to make sure I understand - you want to use the --tlscacert config option for dockerd, but don't have a way to place the CA on the machine via charm config. Does that sum it up?

**Comment by CalvinHartwell** [comment]
> Just want to make sure I understand - you want to use the --tlscacert config option for dockerd, but don't have a way to place the CA on the machine via charm config. Does that sum it up? Nevermind, marcoceppi set me straight.

What exactly did he explain? 🎱 Drop me a mail if you want to ask questions, my IRC connectivity is shady at the moment: calvin.hartwell@canonical.com

**Comment by marcoceppi** [comment]
Just want to make sure we understand - You have a private registry with a self-signed certificate, or a certificate that doesn't have a public CA trust already present on the host. So you want to be able to drop that public CA cert on host so that docker can validate the certs of that registry?

**Comment by CalvinHartwell** [comment]
> Just want to make sure we understand - You have a private registry with a self-signed certificate, or a certificate that doesn't have a public CA trust already present on the host. So you want to be able to drop that public CA cert on host so that docker can validate the certs of that registry?

Exactly: we want to be able to pass the public certificate for the CA which signed the private registry certificate so the client can verify it (*note, this could be a chain of public certificates). I assume this means the certificate will be added to a truststore on the machine or in some folder which acts as a trust store for the docker daemon. 

* Note: Most big companies will use an intermediate CA (which sign on behalf of the root CA. This usually means that a bundle of certificates are added together, because a client can only verify that a certificate has been signed by an intermediate CA by performing a check with the public root CA certificate. We would also expect this to be correctly handled and working.

**Comment by marcoceppi** [comment]
Yeah, we may honestly just produce a truststore subordinate - this case isn't unique to docker and I imagine a lot of charms could benefit from a finer granularity of public trust ring management.

The biggest problem is how coarse the Juju config for this. Odds are you'll just be base64 encoding a blob of public CA certs that the charm then unpacks and updates appropriately.

**Comment by tvansteenburgh** [comment]
1. We should not try to generalize the solution too early. Let's solve the problem at hand and generalize later if appropriate. In other words, let's fix layer-docker and revisit the subordinate idea later.
2. Why not use resources instead of config?

**Comment by marcoceppi** [comment]
@CalvinHartwell A better option, if you have juju 2.3 would be to use cloud-init (https://cloudinit.readthedocs.io/en/latest/topics/examples.html#configure-an-instances-trusted-ca-certificates) as a model-default. Thereby all machines would have that cert loaded. If that works for you then you have an immediate fix for now while the engineering team looks into other alternatives.

```
juju model-defaults cloudinit-userdata="$(cat /path/to/cloud-init.yaml)"
```

This can also be set during bootstrap.

**Comment by marcoceppi** [comment]
@tvansteenburgh 

1. The process for updating ca-certificates is a global problem, not a docker problem. The amount of effort it will take to engineer for layer-docker is the same it would take to do it right. https://askubuntu.com/a/649463/41

2. Because resources are a PITA to make repeatable. Public CA certs are configuration, not payloads. Also, as config you can have a bundle and it will just work. public ca files aren't secret and won't be problematic in doing so.

Finally, I think the model-config + cloud-init might negate needing to have any engineering effort whatsoever.

**Comment by wwwtyro** [comment]
@CalvinHartwell Are you able to comment on whether or not the cloudinit solution will work for you?

**Comment by CalvinHartwell** [comment]
> @CalvinHartwell Are you able to comment on whether or not the cloudinit solution will work for you?

@wwwtyro I would need to test it to be sure. I'd much prefer this to be an option for Juju just like the docker-opts or other options.

**Comment by marcoceppi** [comment]
Well, this is a Juju option, I believe you mean charm level option, which
is fine.

**Comment by wwwtyro** [comment]
There appears to be a problem with using cloudinit-userdata: https://bugs.launchpad.net/juju/+bug/1748283

**Comment by wwwtyro** [comment]
aaaand the fix is already in: https://github.com/juju/juju/pull/8354

Thanks, @hmlanigan

**Comment by wwwtyro** [comment]
This fix is in 2.3.3. I think I'll close this issue for now, but please reopen if you see a need.
